rust/windows-client/src-tauri/src/client/auth.rs

```diff
@@ -0,0 +1,353 @@
+    /// May return Ok(None) if we're already signed in
+    pub fn start_sign_in(&mut self) -> Result<Option<&Request>> {
+        self.sign_out()?;
+        self.state = State::NeedResponse(Request {
+            nonce: generate_nonce(),
+            state: generate_nonce(),
+        });
+        Ok(Some(self.ongoing_request()?))
+    }
+
+    /// Complete an ongoing sign-in flow using parameters from a deep link
+    ///
+    /// Returns a valid token.
+    /// Performs I/O.
+    ///
+    /// Errors if we don't have any ongoing flow, or if the response is invalid
+    pub fn handle_response(&mut self, resp: Response) -> Result<SecretString> {
+        let req = self.ongoing_request()?;
+
+        if !secure_equality(&resp.state, &req.state) {
+            self.sign_out()?;
+            return Err(Error::StatesDontMatch);
+        }
+
+        let token = format!(
+            "{}{}",
+            req.nonce.expose_secret(),
+            resp.fragment.expose_secret()
+        );
+        self.keyring_entry()?.set_password(&token)?;
+        self.state = State::SignedIn(Session {
+            actor_name: resp.actor_name,
+        });
+        Ok(SecretString::from(token))
+    }
+
+    /// Returns the token if we are signed in
+    ///
+    /// This may always make syscalls, but it should be fast enough for normal use.
+    pub fn token(&self) -> Result<Option<SecretString>> {
+        match self.state {
+            State::SignedIn(_) => {}
+            _ => return Ok(None),
+        }
+
+        self.get_token_from_disk()
+    }
+
+    /// Retrieves the token from disk regardless of in-memory state
+    ///
+    /// Performs I/O
+    fn get_token_from_disk(&self) -> Result<Option<SecretString>> {
+        Ok(match self.keyring_entry()?.get_password() {
+            Err(keyring::Error::NoEntry) => None,
+            Ok(token) => Some(SecretString::from(token)),
+            Err(e) => Err(e)?,
+        })
+    }
+
+    /// Returns an Entry into the OS' credential manager
+    ///
+    /// Anything you do in there is technically blocking I/O.
+    fn keyring_entry(&self) -> Result<keyring::Entry> {
+        Ok(keyring::Entry::new_with_target(self.keyring_key, "", "")?)
+    }
+
+    fn ongoing_request(&self) -> Result<&Request> {
+        match &self.state {
+            State::NeedResponse(x) => Ok(x),
+            _ => Err(Error::NoInflightRequest),
+        }
+    }
+}
+
+/// Generates a random nonce using a CSPRNG, then returns it as hexadecimal
+fn generate_nonce() -> SecretString {
+    let mut buf = [0u8; NONCE_LENGTH];
+    // rand's thread-local RNG is said to be cryptographically secure here: https://docs.rs/rand/latest/rand/rngs/struct.ThreadRng.html
+    thread_rng().fill_bytes(&mut buf);
+
+    // Make sure it's not somehow all still zeroes.
+    assert_ne!(buf, [0u8; NONCE_LENGTH]);
+    hex::encode(buf).into()
+}
+
+/// Checks if two byte strings are equal in constant-time.
+/// May not be constant-time if the lengths differ.
```

Comment:
AFAICT it should always be constant in time, right?

Reply:
The `subtle` docs said it can short-circuit if the lengths differ https://docs.rs/subtle/2.5.0/subtle/trait.ConstantTimeEq.html#impl-ConstantTimeEq-for-%5BT%5D